intel: ssp: fix RX fifo drain logic

kv2019i opened the pull request:
Separate the RX drain functions for start and stop cases. For startup, do not add any delays but read the full fifo out and ensure the dma request line (SSSR.RFS) is cleared before new capture stream is started.

Link: https://github.com/thesofproject/sof/issues/7548
Fixes: 4a4d8d261907 ("intel: ssp: drain RX fifo when starting")

btian1 commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -98,10 +119,10 @@ static void ssp_empty_rx_fifo(struct dai *dai)
-				ssp_read(dai, SSDR);
-		}
+#if CONFIG_DMA_SUSPEND_DRAIN
+		for (i = 0; i < entries + 1; i++)
```
can remove 106 - 108, and use:
for ( int i = 0; i < entries + 1; i++)

kv2019i replied:
Yeah, that would be nicer, but for now, we've stick to Linux kernel style and declare variable at start. I know we start to have some C++ inspired code that is not following this anymore...

btian1 reviewed (commented):
this is a follow up #7577, one more question is: the change only have impact on BT or all SSP related operation? should be all, right?

ujfalusi commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -67,23 +67,44 @@ static void ssp_empty_tx_fifo(struct dai *dai)
+	while ((ssp_read(dai, SSSR) & SSSR_RNE) && retry--) {
+		entries = SSCR3_RFL_VAL(ssp_read(dai, SSCR3));
+
+		/* let DMA consume data or read RX FIFO directly */
```
the comment is not correct.

ujfalusi commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -67,23 +67,44 @@ static void ssp_empty_tx_fifo(struct dai *dai)
+		for (i = 0; i < entries + 1; i++)
+			ssp_read(dai, SSDR);
+
+		entries = SSCR3_RFL_VAL(ssp_read(dai, SSCR3));
```
no need to re-read the entries here.

ujfalusi commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -67,23 +67,44 @@ static void ssp_empty_tx_fifo(struct dai *dai)
+	}
+
+	/* clear interrupt */
+	ssp_update_bits(dai, SSSR, SSSR_ROR, SSSR_ROR);
```
Clear Receive FIFO overrun

ujfalusi commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -67,23 +67,44 @@ static void ssp_empty_tx_fifo(struct dai *dai)
+}
+
 /* empty SSP receive FIFO */
 static void ssp_empty_rx_fifo(struct dai *dai)
```
if the other incarnation is `ssp_empty_rx_fifo_for_start()` then this is `ssp_empty_rx_fifo_for_stop()`?

ujfalusi commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -98,10 +119,10 @@ static void ssp_empty_rx_fifo(struct dai *dai)
+#if CONFIG_DMA_SUSPEND_DRAIN
+		for (i = 0; i < entries + 1; i++)
+			ssp_read(dai, SSDR);
+#endif
```
I'm not certain about this. Even if the DMA is active (not disabled) but the RX FIFO does not have trigger amount of data then the DMA will not clear the FIFO, so we just spin w/o any reason.

kv2019i replied:
@ujfalusi Ack, so this will not drain the samples that fall below the DMA trigger threshold and the function will timeout (which at 8kHz is a long timeout). OTOH, in old code fifo is emptied directly and DMA is left without trailing data, so not great either.

Vamshigopal commented:
@kv2019i  i've tested the patch with noraml scenario and with no-codec tplg with 8k rate. Issue is not reproducible.

kv2019i commented:
Excellent. thanks @Vamshigopal ! @ujfalusi is working on a revised patch we are currently stress testing on multiple devices. This builds on this same fix, but addresses some of the issues raised in this review. We'll share this later for review.

lyakh commented on src/drivers/intel/ssp/ssp.c:
```diff
@@ -67,23 +67,44 @@ static void ssp_empty_tx_fifo(struct dai *dai)
+/*
+ * Empty SSP receive FIFO before DMA startup and
+ * ensure the DMA request is not already raised when
+ * DMA is started. If TX is not active, this is a no-op.
```
I'm confused a bit: isn't receive FIFO usually the FIFO where data arrives from the interface, i.e. for capture? Why is it talking about TX then?

Vamshigopal commented:
I've encountered a strange fimding while testing this patch.

Issue: No sound from playback (BT output ) 
Issue scenario:

steps:
1. start yt playback from Btoutput 
2. start recording from Btinput
3. stop recording from btinput
4. repeat 2,3 for couple of times
5. reboot
6. start yt playback from btoutput

In step 6 we dont see any sound from btoutput.

kv2019i commented:
Closing, let's go with #7717  instead.
